File: leetcode-04/solutions/cpp/0037.cpp
class Solution {
 public:
  void solveSudoku(vector<vector<char>>& board) { dfs(0, board); }

 private:
  bool dfs(int s, vector<vector<char>>& board) {
    if (s == 81) return true;

    int i = s / 9;
    int j = s % 9;

    if (board[i][j] != '.') return dfs(s + 1, board);

    for (char c = '1'; c <= '9'; c++)
      if (isValid(i, j, c, board)) {
        board[i][j] = c;
        if (dfs(s + 1, board)) return true;
        board[i][j] = '.';
      }

    return false;
  }

  bool isValid(int row, int col, char c, vector<vector<char>>& board) {
    for (int i = 0; i < 9; i++) {
      if (board[i][col] != '.' && board[i][col] == c) return false;
      if (board[row][i] != '.' && board[row][i] == c) return false;
      if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] != '.' &&
          board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c)
        return false;
    }
    return true;
  }
};

File: leetcode-04/solutions/python3/0037.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        self.dfs(0, board)

    def dfs(self, s: int, board: List[List[str]]) -> bool:
        if s == 81:
            return True

        i = s // 9
        j = s % 9

        if board[i][j] != '.':
            return self.dfs(s + 1, board)

        for c in range(1, 10):
            if self.isValid(i, j, str(c), board):
                board[i][j] = str(c)
                if self.dfs(s + 1, board):
                    return True
                board[i][j] = '.'

        return False

    def isValid(self, row: int, col: int, c: str, board: List[List[str]]) -> bool:
        for i in range(9):
            if board[i][col] != '.' and board[i][col] == c:
                return False
            if board[row][i] != '.' and board[row][i] == c:
                return False
            if board[3 * (row // 3) + i // 3][3 * (col // 3) + i % 3] != '.' and \
                    board[3 * (row // 3) + i // 3][3 * (col // 3) + i % 3] == c:
                return False
        return True
